Handle strict evaluation of Sized types

leonschoorl opened the pull request.

christiaanb commented on clash-lib/src/Clash/Normalize/Transformations.hs:
```diff
@@ -557,11 +559,20 @@ caseCast _ e = return e
   changed $ Letrec $ bind binds (Cast body ty1 ty2)
 letCast _ e = return e
 
+-- | Is this a 'Cast' that you be pushed down?
```
Elaborate why some casts should or shouldn't be pushed down

christiaanb commented on clash-lib/src/Clash/Normalize/Transformations.hs:
```diff
@@ -557,11 +559,20 @@ caseCast _ e = return e
+shouldPushDown _ = False
+
+argCastSpec :: NormRewrite
+argCastSpec ctx e@(App _ c@(Cast _ _ _)) | shouldPushDown c = specializeNorm ctx e
```
Shouldn't we ensure that the cast is work-free?? so let-bind any work before we specialise.

christiaanb commented on clash-lib/src/Clash/Core/Type.hs:
```diff
@@ -513,3 +515,33 @@ isIntegerTy :: Type -> Bool
+      | name2String tcNm == "Clash.Signal.Internal.Signal"
+      , [_,elTy] <- args
+      -> go elTy
+      | name2String tcNm == "Clash.Sized.Internal.BitVector.BitVector" ||
```
Add a comment that you are looking through newtypes, but that clash' primitive number types are implemented as newtypes as well and you don't want to normalize them away.

christiaanb commented on clash-lib/src/Clash/Core/Evaluator.hs:
```diff
@@ -297,6 +301,10 @@ unwind eval gbl tcm h k v = do
     Instantiate ty               -> return (instantiate h k' v ty)
     PrimApply nm ty tys vals tms -> primop eval gbl tcm h k' nm ty tys vals v tms
     Scrutinise _ alts            -> return (scrutinise h k' v alts)
+    CastTo ty1 ty2               -> return (castTo h k' v ty1 ty2)
```
This is wrong.... what it should be doing e.g. in a cast from `Integer` to `Signed`, is to add the `Signed.fromInteger#` primitive! The reason being, that primitives operating on `Signed` are expecting to see this primitive, they're not expecting to see a cast.

leonschoorl commented on clash-lib/src/Clash/Normalize/Transformations.hs:
```diff
@@ -538,6 +542,58 @@ bindConstantVar = inlineBinders test
+-- @
+--   (cast :: b -> a) $ (cast :: a -> b) x   ==> x
+-- @
+eliminateCastCast :: NormRewrite
```
It's good to check tyA and tyB.
But we need to do type normalization before comparing them, because they could be for example: `Signed 7` and `Signed (6+1)`.

I'm changing the case where tyA and tyC don't match, to produce an error.
Introducing a new cast here will produce a cast that can never be eliminated.
And then we will get an error during netlist generation that there was still some cast left over, but at that point we won't know where it came from.

leonschoorl replied:
Also in the test case that I have this doesn't actually provide any source location.

leonschoorl commented:
I addressed all your comments in a series of patches in the [seq-on-sized-types-reviewed](https://github.com/leonschoorl/clash-compiler/commits/seq-on-sized-types-reviewed) branch.
Then I merged those patches into the original patches and force-pushed the result overwriting the  [seq-on-sized-types](https://github.com/leonschoorl/clash-compiler/commits/seq-on-sized-types).

But the Github UI seems very confused by this.

christiaanb commented on clash-lib/src/Clash/Normalize/Transformations.hs:
```diff
@@ -538,6 +548,116 @@ bindConstantVar = inlineBinders test
+-- @
+eliminateCastCast :: NormRewrite
+eliminateCastCast _ c@(Cast (Cast e tyA tyB) tyB' tyC)
+  | tyB == tyB' = do
```
I think we want to compare normalised types here as well, no?

leonschoorl replied:
Well `tyB` and `tyB'` should always be `Integer`.
But I'll add a check just to be sure.

christiaanb commented on clash-lib/src/Clash/Normalize/Transformations.hs:
```diff
@@ -538,6 +548,116 @@ bindConstantVar = inlineBinders test
+  | otherwise = throwError
+  where throwError = do
+          (nm,sp) <- Lens.use curFun
+          throw (ClashException sp ($(curLoc) ++ showDoc nm ++ ": Found 2 nested casts whose types don't line up:\n" ++ showDoc c) Nothing)
```
Can we try to stay closer to a 80-column window in our formatting so that code is easier to read on smaller terminals?

christiaanb commented on clash-lib/src/Clash/Normalize/Transformations.hs:
```diff
@@ -538,6 +548,116 @@ bindConstantVar = inlineBinders test
+
+splitCastWork _ e = return e
+
+
```
These extra newlines aren't needed
